README.md
# Front Authorization Handler

A Javascript frontend authorization tokens handler that works with JWT to check expiration and token transactions with the server.
 
Handle authorization tokens as sessions, based on `accessToken` and `refreshToken`, that will save it in local or session storage, and create an automatic renew when tokens expire and can integrate with request interceptors.

 ![license](https://camo.githubusercontent.com/76d5d2b7f6cb797adf6c30fafa4a2cb2f4390155/68747470733a2f2f696d672e736869656c64732e696f2f6769746875622f6c6963656e73652f6d61746961736c6f70657a642f584465627567676572) ![size](https://img.shields.io/bundlephobia/min/@videsk/front-auth-handler) ![sizesrc](https://img.shields.io/github/size/videsk/front-auth-handler/index.js) ![issues](https://img.shields.io/github/issues-raw/videsk/front-auth-handler) ![rank](https://img.shields.io/librariesio/sourcerank/npm/@videsk/front-auth-handler) ![version](https://img.shields.io/npm/v/@videsk/front-auth-handler) ![downloads](https://img.shields.io/npm/dt/@videsk/front-auth-handler) [![DeepScan grade](https://deepscan.io/api/teams/7725/projects/9797/branches/130453/badge/grade.svg)](https://deepscan.io/dashboard#view=project&tid=7725&pid=9797&bid=130453)

```
Use the latest version! Previous version < 3.0.0 are insecure
and has not been tested.
```

## Features

* Automatic save tokens in session or local storage
* Automatic storage for remember a session or not
* Expiration observer
* Server side validation
* Automatically and manually renew the access token with the refresh token
* Events for expiration, renewed, empty and errors

## Installation

From NPM

`npm i @videsk/front-auth-handler`

Self hosted

`<script src="/dist/web-auth.min.js"></script>`
 
## How to use
For start you need instance `WebAuth`.

```js
const auth = new WebAuth(options);
```

### Options

The class receive one argument as `object`. That object contains the following keys:

```js
const options = { keys, config };
```

**keys**

This key, have the "key name" that will store in local or session storage. By default, are `auth-key` and `auth-key-refresh`, for access and refresh token respectively.

```js
const keys = {
    accessToken: 'auth-key',
    refreshToken: 'auth-key-refresh'
};
```

**config**

In case of config, this key has the major of configuration of WebAuth, so we recommend go to `index.js` [file](https://github.com/videsk/front-auth-handler/blob/master/index.js) and copy the default object.

In previous versions, use server side validation was optional, now from version 3.0.0 is mandatory, so this decision was took by security concerns. Please not send PR or issue requesting that.

### Start to handle

After you pass the options that is optional (but you will work in `http://localhost:3000`, by default) can start to handle with the following code:

```js
// Set empty, auto handle existing saved tokens
auth.set();
// Set with accessToken only
auth.set(accessToken);
// Set with accessToken and refreshToken
auth.set(accessToken, refreshToken);
// Also can set if you want "remember session"
auth.set(accessToken, refreshToken, true);
```

This automatically start to observe the expiration of access and/or refresh token. Also check validity of access token by the server and when expire automatically will try to get a new one, only if you set a refresh token.

### Force renew

This property allows you to force renew the access token with the refresh token, so this is relevant when you integrate with interceptors. In case a request returns `401 Unauthorized` (typically token expire or blacklisted) you can force renew.

```js
auth.renew();
```

This is very useful not only when user do a request, seconds before the observer can check that access token was expired, also when access and/or refresh token was invalidated by the server in a blacklist by security reasons (ex. password reset).

So, is important use it directly in your interceptors when you know that access token was expired, blacklisted or invalidated by the server.

### Events

WebAuth from version 3.0.0 was added events, can help to know when access and refresh token was expired, renewed or something not works.

The available events are:

```js
const events = {
    expired: () => {}, 
    error: () => {}, 
    renewed: () => {}, 
    empty: () => {},
    load: () => {},
};
```

To set them, you can do it with this elegant way:

```js
auth.on('name-event', callback);

// Example
auth.on('expired', function () {
    // Do something
});

// Or less elegant
auth.events.expired = () => {
    // Do something
};
```
Remember that both method will override the default empty function. So, define a callback per event only one time on your code.

In case of `expired` event, returns the name of token was expired that will be `accessToken` or `refreshToken`. So, you can difference like this:

```js
auth.on('expired', function (tokenName) {
    if (tokenName === 'accessToken') return; // Set function when not use refreshToken
    // Is refreshToken
    logoutUser();
});
```

The `load` events is useful to set in the base of HTML or template app to know when the accessToken is valid or was renewed. Example:

```js
auth.on('load', function() {
    // The accessToken and/or refreshToken are valid
    // Or accessToken is expired, WebAuth will try to renew, then if the renovation is successful the event will be triggered
});
```

So, is recommended to add the `load` event to the base of the app. With that, you can ensure that `load` event will be triggered only if accessToken and/or refreshToken are been valid. Inclusive if was valid from the first time the app was loaded or need to be renewed. Both cases ensure that the app can load with valid accessToken.

### Stop and clean

This two properties allows you to `clean` tokens from storage and `stop` the observer. The `stop` also clean the tokens, so use stop when you want to reactivate observer manually with `.set()`.

```js
// Only clean tokens
auth.clean();
// Stop all and clean tokens
auth.stop();
```

We recommend to use `.stop()` when the user logout from the web app.

### Server side validation

When you start `WebAuth`, try to check validity to the endpoint, in case the server returns an expired response, `WebAuth` will try to get a new one with the refreshToken.

In the `config` key, you will see that the two endpoints have `status` and `attempts`. These keys help to handle the response of the server, so in case of `status = { ok: 200, expired: 401 }` tells to `WebAuth` when the server returns a new one or validate correctly the accessToken. And in case of `attempts` in when the server returns other code that is not specified on the `status` and give the possibility to try again, also when the user does not have Internet connection.

When `WebAuth` detects issues trying to get a response of the server, automatically start to try N times you set on `attempts`, as a number of intervals and factor. That means:

```js
setTimeout(() => recursive(attempts + 1), 1000 * attempts + 1);

// Example with the 3 attempts by default

// First attempt
setTimeout(() => recursive(1), 1000);
// Second attempt
setTimeout(() => recursive(2), 2000);
// Third attempt
setTimeout(() => recursive(3), 3000);
// ...

// So in a period of 6 seconds WebAuth will try to get a ok or expired response from server with 3 attempts
```

In case exceed the attempts the event error will throw, and the tokens will remove from the session or local storage. That behavior is by security reasons, to avoid store tokens without server validation, so you can override calling to method `stop()`.

This is really helpful when user lose connection, also in that cases you can complement with manually Internet connection check, so you can call `stop()` method to avoid `WebAuth` remove tokens. 

## Lifecycle

This is the lifecycle of `WebAuth`, when use accessToken and refreshToken.

```shell
    Instance WebAuth
            ↓
set(accessToken, refreshToken) → observer start → (if empty, no tokens) → fire empty()
            ↓
server side accessToken validate
            ↓                           
    (accessToken expire) → fire expire('accessToken')
            ↓                       
    renew automatically ⇿ (if fails) ← try renew x times → fire error(error)
            ↓
        save tokens → observer start → fire renewed()
            ↓
    refreshToken expire
            ↓
fire expired('refreshToken')
            ↓
  end (here request login)
```

If you don't use a refreshToken, the lifecycle will be like this:

```shell
    Instance WebAuth
            ↓
server side accessToken validate
            ↓
set(accessToken, refreshToken) → observer start → (if empty, no tokens) → fire empty()
            ↓
(accessToken expire) → fire expire('accessToken')
            ↓
           end
```

# Debugging

We recommend to use event `error` to check all issues with server side validations. This library was tested so should not throw unhandled errors related to observer.

```js
// Development environment
auth.on('error', function (error) {
    debugger;
    console.log(error);
});

// Production
auth.on('error', function (error) {
    // Here can integrate with error monitoring like Sentry
});
```

We strongly recommend use `error` event with error monitoring like Sentry, Bugsnag, LogRocket, etc. The cases when error event should be fired are a malformed JWT and server error response.

# Testing

This library was tested with `Mocha`, `chai` and `chai-http`. Also was created a polyfill of `window` to test with `localStorage` and `sessionStorage` in Node, [check here](https://github.com/videsk/window-node-polyfill).

For coverage was used `nyc`.

# Changelog

See changelog [here](https://github.com/videsk/front-auth-handler/blob/master/CHANGELOG.MD).

# License

This library was developed by [Videsk](https://videsk.io) with ♥ license LGPL-2.1. 


CHANGELOG.MD
# Changelog

The versioning is based on a semantic version (semver).

### v3.2.1

- [NEW] Fixed `_stop` key after execute `stop()` method.

### v3.2.0

- [NEW] HTTP code status can be number or array with numbers.

### v3.1.2

- [FIX] Added to `stop` method set initial state, `_load` property to false

### v3.1.1

- [FIX] Change lifecycle `load` event

### v3.1.0

- [NEW] Load event

### v.3.0.6

- [FIX] Stable version

### v3.0.0 [Break change]

**CAUTION**: this version will crash your current project, check [README](https://github.com/videsk/front-auth-handler#readme) to do changes correctly.

This version is focused on security tokens handle and integration with request interceptors to force renew the access token if JWT could be invalid by others factors like blacklist. 

- [NEW] Completely rebased
- [NEW] Events expired, renewed, empty and error
- [NEW] Force renew
- [NEW] Mocha test
- [NEW] Coverage test
- [NEW] Examples

### v2.4.0

- [FIX] Completely refactored (5Kb less)
- [NEW] jsDocs added to source code
- [NEW] Dynamic Body parser added
- [NEW] Max attempts added
- [TODO] Add max attempts to internet issues (v3.0.0)
- [TODO] Add proxy handle request failed by 401 (v3.0.0)
- [TODO] Update README.md (v3.0.0)
- [TODO] Add examples how to use (v3.0.0)
- [TODO] Add tests (v3.0.0)

### v2.3.0

- [FIX] Fixed infinite interval when error code is different of custom code

### v2.0.7

- [FIX] Added changelog to readme

### v2.0.6

- [FIX] Auto detect storage or manually for set boolean in remember.
- [NEW] Added changelog

### v2.0.5

- [FIX] Corrected some words in readme

### v2.0.4

- [FIX] Corrected some words in readme

### v2.0.0 - v2.0.3

- [NEW] Added HTTP tokens validation and implementation of refresh token for auto-generate a new access token via HTTP.


**Previous versions are deprecated, we recommend not use.**


test/server.js
const express = require("express");
const app = express();
const server = require("http").createServer(app);
const jwt = require("jsonwebtoken");
const authorization = require('auth-header');
const bodyParser = require('body-parser');

app.use(bodyParser.json());

const jwtOptionsAccessToken = {
  secret: "b98U897b6r5567y89NM0m",
  expiresIn: "3s"
};

const jwtOptionsRefreshToken = {
  secret: "zny283en2837eze23e23e",
  expiresIn: "10s"
};

// Set node app trust in proxy
app.set("trust proxy", true);

// Create and send a new access token in endpoint /access-token
app.get("/login", function (req, res) {
  // Create JWT
  const accessToken = jwt.sign(
      { iat: Math.floor(new Date().getTime() / 1000) },
      jwtOptionsAccessToken.secret,
      { expiresIn: jwtOptionsAccessToken.expiresIn }
  );
  const refreshToken = jwt.sign(
      { iat: Math.floor(new Date().getTime() / 1000) },
      jwtOptionsRefreshToken.secret,
      { expiresIn: jwtOptionsRefreshToken.expiresIn }
  );
  // Set response have JSON format
  res.header("Content-Type", "application/json");
  // Send response
  res.json({
    accessToken,
    refreshToken
  });
});

app.get("/login-expired", function (req, res) {
  // Create JWT
  const accessToken = jwt.sign(
      { iat: Math.floor(new Date().getTime() / 1000) },
      jwtOptionsAccessToken.secret,
      { expiresIn: 1 }
  );
  const refreshToken = jwt.sign(
      { iat: Math.floor(new Date().getTime() / 1000) },
      jwtOptionsRefreshToken.secret,
      { expiresIn: jwtOptionsRefreshToken.expiresIn }
  );
  // Set response have JSON format
  res.header("Content-Type", "application/json");
  // Send response
  res.json({
    accessToken,
    refreshToken
  });
});

app.get("/check-token", function (req, res) {
  try {
    const {token: accessToken} = authorization.parse(req.header('Authorization'));
    // Validate JWT
    jwt.verify(accessToken, jwtOptionsAccessToken.secret);
    // Set response have JSON format
    res.header("Content-Type", "application/json");
    // Send response
    res.json({ message: 'The accessToken is valid.' });
  } catch (e) {
    res.status(401);
    res.json({ message: e.message });
  }
});

app.post("/refresh-token", function (req, res) {
  try {
    const {refreshToken} = req.body;
    jwt.verify(refreshToken, jwtOptionsRefreshToken.secret);

    const accessToken = jwt.sign(
        { iat: Math.floor(new Date().getTime() / 1000) },
        jwtOptionsAccessToken.secret,
        { expiresIn: jwtOptionsAccessToken.expiresIn }
    );
    // Set response have JSON format
    res.header("Content-Type", "application/json");
    res.status(201);
    // Send response
    res.json({ accessToken });
  } catch (e) {
    res.status(401);
    res.json({ message: e.message });
  }
});

// Handle exceptions
process.on("uncaughtException", function (err) {});
// Handle rejections
process.on("unhandledRejection", function (reason, p) {});

module.exports = {
  jwtOptionsAccessToken,
  jwtOptionsRefreshToken,
  server,
};


test/options.js
const options = {
    endpoints: {
        hostname: 'http://localhost:3000/',
        accessToken: {
            url: 'check-token',
            method: 'GET',
            authorizationType: 'Bearer',
            headers: {
                'Content-Type': 'application/json'
            },
            body: {},
            keys: {
                accessToken: 'accessToken',
                refreshToken: 'refreshToken',
            },
            status: {
                expired: 401,
                ok: 200,
            },
            attempts: 3
        },
        refreshToken: {
            url: 'refresh-token',
            method: 'POST',
            authorizationType: 'Bearer',
            headers: {
                'Content-Type': 'application/json'
            },
            body: {},
            keys: {
                accessToken: 'accessToken',
                refreshToken: 'refreshToken',
            },
            status: {
                expired: 401,
                ok: 201,
            },
            attempts: 3
        },
    },
};

module.exports = options;


index.js
const { version } = require('./package.json');

class WebAuth {

    constructor(options = {}) {
        const {
            keys = { accessToken: 'auth-key', refreshToken: 'auth-key-refresh' },
            config = {
                endpoints: {
                    hostname: 'http://localhost:3000/',
                    accessToken: {
                        url: 'check-token',
                        method: 'GET',
                        authorizationType: 'Bearer',
                        headers: {
                            'Content-Type': 'application/json'
                        },
                        body: {},
                        keys: {
                            accessToken: 'accessToken',
                            refreshToken: 'refreshToken',
                        },
                        status: {
                            expired: 401,
                            ok: 200,
                        },
                        attempts: 3
                    },
                    refreshToken: {
                        url: 'refresh-token',
                        method: 'POST',
                        authorizationType: 'Bearer',
                        headers: {
                            'Content-Type': 'application/json'
                        },
                        body: {},
                        keys: {
                            accessToken: 'accessToken',
                            refreshToken: 'refreshToken',
                        },
                        status: {
                            expired: 401,
                            ok: 201,
                        },
                        attempts: 3
                    },
                },
            },
        } = options;

        this.keys = keys;
        this.events = {
            expired: () => {},
            error: () => {},
            renewed: () => {},
            empty: () => {},
            load: () => {},
        };
        this.config = config;
        this.storage = (window.localStorage.getItem(this.keys.accessToken)) ? 'localStorage' : 'sessionStorage';

        this._expirationAccessToken = null;
        this._expirationRefreshToken = null;
        this._stop = false;
        this.version = version;
        this._load = false;
    }

    /**
     * Set WebAuth and start observer
     * @param access {String=} - accessToken
     * @param refresh {String=} - refreshToken
     * @param remember {Boolean=} - Save in session or local storage
     * @returns {Promise<*|undefined>}
     */
    async set(access = '', refresh = '', remember) {
        this._stop = false;
        if (typeof remember === 'boolean') this.storage = remember ? 'localStorage' : 'sessionStorage';
        const { accessToken = access, refreshToken = refresh } = this.constructor.getTokens(this.storage, this.keys);
        if (!accessToken) return this.events.empty();
        // Save expiration
        try {
            this._expirationAccessToken = this.constructor.getExpirationToken(accessToken);
            if (refreshToken) this._expirationRefreshToken = this.constructor.getExpirationToken(refreshToken);
        } catch (e) {
            return this.events.error(e);
        }
        this.constructor.saveTokens(this.storage, this.keys, accessToken, refreshToken);

        try {
            await this.askServer();
            return this.observer();
        } catch (e) {
            if (e instanceof Error) return this.events.error(e);
            this.events.expired('accessToken');
            return this.renew();
        }
    }

    /**
     * Set event
     * @param event
     * @param callback
     */
    on(event = '', callback = () => {}) {
        this.events[event] = callback;
    }

    /**
     * Clean tokens from storage
     */
    clean() {
        // Remove from localStorage
        window.localStorage.removeItem(this.keys.accessToken);
        window.localStorage.removeItem(this.keys.refreshToken);
        // Remove from sessionStorage
        window.sessionStorage.removeItem(this.keys.accessToken);
        window.sessionStorage.removeItem(this.keys.refreshToken);
    }

    /**
     * Start expiration tokens observer
     * @returns {Promise<*|undefined>|NodeJS.Timeout|*}
     */
    observer(attempts = 1) {
        if (this._stop || !this._expirationAccessToken) return;
        const isAccessTokenExpired = new Date().getTime() > this._expirationAccessToken;
        if (!isAccessTokenExpired && !this._load) {
            this._load = true;
            this.events.load();
        }
        if (!isAccessTokenExpired) return setTimeout(() => this.observer(), 1000);
        if (attempts < 2) this.events.expired('accessToken');
        if (!this._expirationRefreshToken) return;
        // Try to get new refreshToken
        return this.renew(attempts);
    }

    /**
     * Force renew accessToken
     * @returns {Promise<*|undefined|NodeJS.Timeout>}
     */
    async renew(attempts = 1) {
        // Check refreshToken was not expired
        const isRefreshTokenExpired = new Date().getTime() > this._expirationRefreshToken;
        if (isRefreshTokenExpired) return this.events.expired('refreshToken');

        const { refreshToken: config } = this.config.endpoints;
        const { refreshToken } = this.constructor.getTokens(this.storage, this.keys);
        try {
            const { accessToken } = config.keys;
            const response = await this.askServer('refreshToken');
            this.constructor.saveTokens(this.storage, this.keys, response[accessToken], refreshToken);
            this._expirationAccessToken = this.constructor.getExpirationToken(response[accessToken]);
            this._expirationRefreshToken = this.constructor.getExpirationToken(refreshToken);
            this.events.renewed();
            return this.observer();
        } catch (e) {
            if (!(e instanceof Error)) return; // Only if was expired
            if (attempts >= config.attempts) return this.events.error(e);
            setTimeout(() => this.observer(attempts + 1), 1000 * (attempts + 1));
        }
    }

    /**
     * Send to server request to check or renew access token
     * @param tokenName {String} - Could be accessToken or refreshToken
     * @returns {Promise<any>}
     */
    async askServer(tokenName = 'accessToken') {
        if (this._stop) return;
        const {url, body, headers, method, authorizationType, keys, status} = this.config.endpoints[tokenName];
        // Create payload
        const tokens = this.constructor.getTokens(this.storage, this.keys);
        const xHeaders = Object.assign(headers, {Authorization: `${authorizationType} ${tokens[tokenName]}`});
        const xBody = Object.assign(body, { [keys[tokenName]]: tokens[tokenName] });
        // Send request
        const fullURL = `${this.config.endpoints.hostname}${url}`;
        const payload = { method, headers: xHeaders };
        if (['post', 'patch', 'put'].includes(method.toLowerCase())) payload.body = JSON.stringify(xBody);
        const response = await fetch(fullURL, payload);
        if (response.status === status.ok || (Array.isArray(status.ok) && status.ok.includes(response.status))) return response.json();
        if (response.status !== status.expired && response instanceof Error || (Array.isArray(status.expired) && response.expired.includes(response.status))) throw response;
        // Clean store and fire expired event
        throw this.events.expired(tokenName);
    }

    /**
     * Stop observer and clean tokens
     * @param value
     */
    stop(value = true) {
        this._stop = value;
        this._load = false;
        this.clean();
    }

    /**
     * Get tokens from storage
     * @param storage {String} - Session or local storage
     * @param keys {Object} - Key names of storage
     * @returns {{accessToken: *, refreshToken: *}}
     */
    static getTokens(storage = 'sessionStorage', keys = {}) {
        const accessToken = window[storage].getItem(keys.accessToken);
        const refreshToken = window[storage].getItem(keys.refreshToken);
        return (accessToken || refreshToken) ? { accessToken, refreshToken } : {};
    }

    /**
     * Save tokens in storage
     * @param storage {String} - Name of storage
     * @param keys {Object} - Key names of storage
     * @param accessToken {String} - JWT access
     * @param refreshToken {String} - JWT refresh
     */
    static saveTokens(storage = 'sessionStorage', keys = {}, accessToken = '', refreshToken = '') {
        window[storage].setItem(keys.accessToken, accessToken);
        window[storage].setItem(keys.refreshToken, refreshToken);
    }

    /**
     * Get expiration of JWT
     * @param JWT {String} - Valid JWT
     * @returns {number|number}
     */
    static getExpirationToken(JWT = '') {
        const decoded = JSON.parse(atob(JWT.split('.')[1]));
        if (typeof decoded !== 'object') throw new Error('Invalid JWT, please check.');
        return ('exp' in decoded) ? decoded.exp * 1000 : Infinity;
    }

    /**
     * Get the name of storage based on tokens existence
     * @param key
     * @returns {string} - Key name want search in storage
     */
    static getStorage(key = '') {
        return (window.localStorage.getItem(key)) ? 'localStorage' : 'sessionStorage';
    }
}

if (typeof define === 'function' && define.amd) {
    define('WebAuth', [], function() {
        return WebAuth;
    });
}

if (typeof module !== 'undefined') {
    module.exports = WebAuth;
}
